### PR TITLE
MGMT-2456 no timeout on wrong boot order

### PR DESCRIPTION
**Pull request by RazRegev**

### CODE REVIEW

**Review by filanov — APPROVED** [review]

**Comment by filanov** [comment]
@eranco74 @oshercc do we have installation timeout on a cluster level?

**Comment by filanov** [comment]
I removed LGTM because i saw a possible issue. 
If there are 3 masters two got into a wrong boot order - there will be no timeout for the boot order but the third master - for example bootstrap will fail on waiting for control plane.
```
2020-10-16T13:36:44.250Z Host rahwvm-os3: updated status from "installing-in-progress" to "installing-pending-user-action" (Expected the host to boot from disk, but it booted the installation image - please reboot and fix boot order to boot from disk /dev/sda (60022480542f4515aed02f11b1cb8e18))
2020-10-16T13:36:44.262Z Host 5026121e-373e-43e5-ad6d-ba71d76f749c: registered to cluster
2020-10-16T13:37:07.381Z Host rahwvm-os1: updated status from "installing-in-progress" to "installing-pending-user-action" (Expected the host to boot from disk, but it booted the installation image - please reboot and fix boot order to boot from disk /dev/sda (6002248007b10fd2a345fd9f6d4d6ac9))
2020-10-16T13:37:07.393Z Host 3076627e-79f8-9ca6-dc29-9083618cebe0: registered to cluster
2020-10-16T13:49:11.788Z Failed to register host: cluster cannot accept new hosts in its current state
2020-10-16T14:35:36.651Z Host rahwvm-os2: updated status from "installing-in-progress" to "error" (Host failed to install because its installation stage Waiting for control plane took longer than expected 1h0m0s)
2020-10-16T14:35:54.366Z Updated status of cluster rahwvm-os46 to error
```
In this case this fix will not help and the solution is more complicated.
Even if we have a solution for the server side i'm not sure how to handle installer side - as far as i know installer have a timeout as well and after some time it will fail and return an error 

cc: @eranco74 @ronniel1 @avishayt @romfreiman

**Comment by avishayt** [comment]
Agree.  This is a partial solution because we still need to make sure we don't fail on installer timeout, as @filanov said.

**Comment by filanov** [comment]
OK so there are several cases that it can help with 
1. All masters get a wrong boot order.
2. Only bootstrap getting wrong boot-order.
3. Workers get a wrong boot order. 
So maybe it worth pushing this fix.

File: internal/host/transition.go [hunk 1]
```diff
@@ -429,6 +429,9 @@ func (th *transitionHandler) HasInstallationInProgressTimedOut(sw stateswitch.St
 	if !ok {
 		return false, errors.New("HasInstallationTimedOut incompatible type of StateSwitch")
 	}
+	if funk.Contains(IgnoreTimeoutCheckOnStatuses, sHost.srcState) {
```
Comment (filanov) [inline internal/host/transition.go:432]: please add a comment explaining why it is done.

File: internal/host/host.go [hunk 2]
```diff
@@ -62,6 +62,11 @@ var InstallationProgressTimeout = map[models.HostStage]time.Duration{
 	"DEFAULT":                                   60 * time.Minute,
 }
 
+var IgnoreTimeoutCheckOnStatuses = [...]string{
+	models.HostStatusInstallingPendingUserAction,
```
Comment (filanov) [inline internal/host/host.go:66]: What is the difference between the two?

Comment (eranco74) [reply internal/host/host.go:66]: HostStatusInstallingPendingUserAction - wrong boot order and we need the user to boot that specific host from disk.
HostStatusResettingPendingUserAction - post cluster reset we need the user to reboot all nodes back to discovery mode.

Comment (filanov) [reply internal/host/host.go:66]: OK thanks, anyway this fix is not relevant because we don't have timeout for those states anyway

Comment (eranco74) [reply internal/host/host.go:66]: As I see it. if the cluster has a node in one of these states, installation timeout should be off. 
So while this is under host.go it cares about other hosts in the cluster.

**Comment by eranco74** [comment]
> I removed LGTM because i saw a possible issue.
> If there are 3 masters two got into a wrong boot order - there will be no timeout for the boot order but the third master - for example bootstrap will fail on waiting for control plane.
> 
> ```
> 2020-10-16T13:36:44.250Z Host rahwvm-os3: updated status from "installing-in-progress" to "installing-pending-user-action" (Expected the host to boot from disk, but it booted the installation image - please reboot and fix boot order to boot from disk /dev/sda (60022480542f4515aed02f11b1cb8e18))
> 2020-10-16T13:36:44.262Z Host 5026121e-373e-43e5-ad6d-ba71d76f749c: registered to cluster
> 2020-10-16T13:37:07.381Z Host rahwvm-os1: updated status from "installing-in-progress" to "installing-pending-user-action" (Expected the host to boot from disk, but it booted the installation image - please reboot and fix boot order to boot from disk /dev/sda (6002248007b10fd2a345fd9f6d4d6ac9))
> 2020-10-16T13:37:07.393Z Host 3076627e-79f8-9ca6-dc29-9083618cebe0: registered to cluster
> 2020-10-16T13:49:11.788Z Failed to register host: cluster cannot accept new hosts in its current state
> 2020-10-16T14:35:36.651Z Host rahwvm-os2: updated status from "installing-in-progress" to "error" (Host failed to install because its installation stage Waiting for control plane took longer than expected 1h0m0s)
> 2020-10-16T14:35:54.366Z Updated status of cluster rahwvm-os46 to error
> ```
> 
> In this case this fix will not help and the solution is more complicated.
> Even if we have a solution for the server side i'm not sure how to handle installer side - as far as i know installer have a timeout as well and after some time it will fail and return an error
> 
> cc: @eranco74 @ronniel1 @avishayt @romfreiman
Now that we have timeout in the server side, I think we can remove the timeout in the assisted-installer.

**Comment by filanov** [comment]
It looks like the requirement is to know if any of the masters is in this state then we need to disable timeout in all the other hosts that are currently in installing state

**Comment by filanov** [comment]
After a conversation with @ronniel1 and @eranco74 the suggestion is to have a new cluster status `installing-pending-user-action` that will indicate if there is a wrong boot order issue in the cluster. this way a as part of HasInstallationInProgressTimedOut this state will indicate that we having boot order issue in one of the masters and disable the timeout. 

@avishayt  what do you thing?

**Comment by avishayt** [comment]
> After a conversation with @ronniel1 and @eranco74 the suggestion is to have a new cluster status `installing-pending-user-action` that will indicate if there is a wrong boot order issue in the cluster. this way a as part of HasInstallationInProgressTimedOut this state will indicate that we having boot order issue in one of the masters and disable the timeout.
> 
> @avishayt what do you thing?

Yes, makes sense

File: internal/host/transition.go [hunk 3]
```diff
@@ -436,6 +454,15 @@ func (th *transitionHandler) HasInstallationInProgressTimedOut(sw stateswitch.St
 	return time.Since(time.Time(sHost.host.Progress.StageUpdatedAt)) > maxDuration, nil
 }
 
+func IsInstallationPaused(clusterID strfmt.UUID, db *gorm.DB) (bool, error) {
+	var cluster common.Cluster
+	err := db.Select("status").Take(&cluster, "id = ?", clusterID.String()).Error
+	if err != nil {
+		return false, err
+	}
+	return swag.StringValue(cluster.Status) == models.ClusterStatusPendingForInput, nil
```
Comment (filanov) [inline internal/host/transition.go:463]: the cluster is not pending for input - it should be a new state, please take a look at the comments in the PR.

**Comment by RazRegev** [comment]
will be reopened after adding new cluster status

**Pull request closed without merging**